src/libsyntax/parse/parser.rs

```diff
@@ -5391,14 +5391,10 @@ impl<'a> Parser<'a> {
             }
             let mut err = self.diagnostic().struct_span_err(id_sp,
                 "cannot declare a new module at this location");
-            let this_module = match self.directory.path.file_name() {
-                Some(file_name) => file_name.to_str().unwrap().to_owned(),
-                None => self.root_module_name.as_ref().unwrap().clone(),
-            };
             err.span_note(id_sp,
                           &format!("maybe move this module `{0}` to its own directory \
                                      via `{0}/mod.rs`",
-                                    this_module));
+                                    paths.name));
```

Comment:
This is supposed to be the name of the file currently being parsed, not the name of the erroneous `mod`.

For example, if we are parsing `foo/bar.rs` and we encounter a non-inline module `mod baz;`, we want to suggest moving `foo/bar.rs` to `foo/bar/mod.rs`. That way, `bar` would be a directory owner (c.f. #37602), so `mod baz;` would expect `foo/bar/baz.rs` (or `foo/bar/baz/mod.rs`) instead of being an error.

To get the name of the file currently being parsed, you could use `self.root_module_name` (skipping the note if it is `None`), or even better, remove the `root_module_name` field entirely and instead use `self.sess.codemap().span_to_filename(id_sp)` (skipping the note if `id_sp` is `DUMMY_SP`).